### ISSUE TITLE
Check projects as packages, not individual modules.

### ISSUE THREAD

**Issue by Orangestar12** [AtomLinter/linter-pylint#220]
### tl;dr
When you pass a file into `pylint`, that file will be assumed as the top-level package. When you pass in a string using Python's import notation (`a.specific.module`) it will lint that module relative to the rest of the package. Is there any possible way this method of linting could be implemented in `linter-pylint`?

### the long of it
Assume the following directory structure:
```
Working Directory
| - package
|   | - __init__.py
|   | - foo
|   |   | - __init__.py
|   |   | - bar.py
```

Assume I am working on the `bar` module. Assume the `bar` module starts with this bit of code:
```py
from . import baz
```
This would import from `foo`. (Code for it would be written in `foo`'s `__init__.py`.)

When I check this code with `linter-pylint`, it will result in an error:
```
Attempted relative import beyond top-level package (relative-beyond-top-level)
```

Indeed, this is the case when I check the file using `pylint ./package/foo/bar.py`, but not, however, when I check with `pylint package.foo.bar`. By specifying which modules are parents, the relative imports went through.

I propose a change be made to how `pylint` is invoked by this package to inspect modules as though they were a part of the packages they may or may not reside inside.

Since we're already retrieving the project path for things, we could inspect the file's location relative to the project path, chop the `.py` off the end, replace the slashes with periods, then feed that string into `linter` to be checked with `pylint` (assuming, of course, that's how this works. I don't into any of this. If this package itself runs `pylint`, then it's as simple as running it with that new string.)

I'd assume this would be checked with a setting inside the package (check all projects as packages) or some extreme file checking to see if we're actually messing with a package (something that would take ages to write and require a bit of refactoring, I'd assume.)

**Comment by ddaanet** [comment]
I was unable to reproduce pylint behaviour you describe. Here is what I did:

```
mkdir package
touch package/__init__.py
mkdir package/foo
mkdir package/foo/__init__.py
echo "from . import baz" > package/foo/bar.py
touch package/foo/baz.py
virtualenv testenv # Creates a virtualenv with Python 2.7.10
testenv/bin/pip install pylint # installs pylint 1.7.1
testvenv/bin/pylint package/foo/bar.py
```

The produce output is:
```
No config file found, using default configuration
************* Module package.foo.bar
C:  1, 0: Black listed name "bar" (blacklisted-name)
C:  1, 0: Missing module docstring (missing-docstring)
W:  1, 0: Unused import baz (unused-import)

-------------------------------------
Your code has been rated at -20.00/10
```

Nothing about `relative-beyond-top-level`. Have you tried using the latest pylint release?

**Comment by en-dash** [comment]
I was able reproduce the bug like this:

```
mkdir -p package/foo
touch package/foo/baz.py
echo 'from . import baz' > package/foo/bar.py 
pylint package/foo/bar.py 
```

The output is:
```
************* Module bar
C:  1, 0: Black listed name "bar" (blacklisted-name)
C:  1, 0: Missing module docstring (missing-docstring)
E:  1, 0: Attempted relative import beyond top-level package (relative-beyond-top-level)
W:  1, 0: Unused import baz (unused-import)
```

My pylint version is:
```
$ pylint --version
pylint 1.7.1, 
astroid 1.5.2
Python 3.5.2 (default, Nov 17 2016, 17:05:23) 
[GCC 5.4.0 20160609]
```

Maybe the bug is specific to Python 3?
Also, slight modifications make it disappear (eg. my first attempt with `bar` in `foo`'s `__init__.py` didn't show `relative-beyond-top-level` either).

**Comment by Orangestar12** [comment]
It may very well be a Python 3 error. ~~Or maybe it's because Pylint is stupid. We may never really know.~~ I haven't used virtualenv ever and I haven't written Python 2 code in ages.

For what it's worth, versions.
```
$ pylint --version
pylint 1.7.1, 
astroid 1.5.2
Python 3.6.1 (default, Mar 27 2017, 00:27:06) 
[GCC 6.3.1 20170306]
```

**Comment by ddaanet** [comment]
Still cannot reproduce it with python3.
```
mkdir package
touch package/__init__.py
mkdir package/foo
mkdir package/foo/__init__.py
echo "from . import baz" > package/foo/bar.py
touch package/foo/baz.py
virtualenv --python=python3 testenv
testenv/bin/pip install pylint
testenv/bin/pylint package/foo/bar.py
```
Output
```No config file found, using default configuration
************* Module package.foo.bar
C:  1, 0: Black listed name "bar" (blacklisted-name)
C:  1, 0: Missing module docstring (missing-docstring)
W:  1, 0: Unused import baz (unused-import)

----------------------------------------------------------------------
Your code has been rated at -20.00/10 (previous run: -20.00/10, +0.00)
```
Versions
```
$ testenv/bin/pylint --version
No config file found, using default configuration
pylint 1.7.1, 
astroid 1.5.3
Python 3.6.0 (default, Mar  3 2017, 13:24:58) 
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]
```

Cannot reproduce it either with `astroid==1.5.2`.

**Comment by ddaanet** [comment]
Too long without feedback.

In may daily use with recent pylint I do not have this problem.

Closing this issue.
